src/library/searchquery.h

```diff
@@ -9,60 +11,63 @@
 
 #include "trackinfoobject.h"
 #include "proto/keys.pb.h"
+#include "util/assert.h"
+#include "util/memory.h"
 
 QVariant getTrackValueForColumn(const TrackPointer& pTrack, const QString& column);
 
 class QueryNode {
   public:
-    QueryNode() {}
+    QueryNode(const QueryNode&) = delete; // prevent copying
     virtual ~QueryNode() {}
 
     virtual bool match(const TrackPointer& pTrack) const = 0;
     virtual QString toSql() const = 0;
+
+  protected:
+    QueryNode() {}
+
+    static QString concatSqlClauses(const QStringList& sqlClauses, const QString& sqlConcatOp);
 };
 
 class GroupNode : public QueryNode {
   public:
-    GroupNode() {}
-    virtual ~GroupNode() {
-        while (!m_nodes.empty()) {
-            delete m_nodes.takeLast();
-        }
-    }
-
-    void addNode(QueryNode* pNode) {
-        m_nodes.append(pNode);
+    void addNode(std::unique_ptr<QueryNode> pNode) {
+        DEBUG_ASSERT(pNode);
+        m_nodes.push_back(std::move(pNode));
     }
 
   protected:
-    QList<QueryNode*> m_nodes;
+    // NOTE(uklotzde): std::vector is more suitable (efficiency)
+    // than a QList for a private member. And QList from Qt 4
```

Comment:
I assumed we switched also to Qt5 now. AFAIK we chose that 1.12 will be the last Qt4 release